File: frontend/lib/utils/validators.dart
/// Shared form field validators.
///
/// Each method returns a [FormFieldValidator] — a function `(String?) -> String?`
/// where `null` means valid. Compose them with [all] when multiple rules apply.
library;

typedef Validator = String? Function(String?);

/// Runs [validators] in order and returns the first error, or null if all pass.
Validator all(List<Validator> validators) {
  return (value) {
    for (final v in validators) {
      final error = v(value);
      if (error != null) return error;
    }
    return null;
  };
}

/// Field must be non-empty.
Validator required([String message = 'Required']) {
  return (v) => (v == null || v.trim().isEmpty) ? message : null;
}

/// Value must not exceed [max] characters (trims first).
Validator maxLength(int max) {
  return (v) {
    if (v == null) return null;
    if (v.trim().length > max) return 'Max $max characters';
    return null;
  };
}

/// Value must be at least [min] characters if non-empty (trims first).
Validator minLength(int min, [String? message]) {
  return (v) {
    if (v == null || v.trim().isEmpty) return null;
    if (v.trim().length < min) {
      return message ?? 'At least $min characters required';
    }
    return null;
  };
}

/// Unicode letters, combining marks, apostrophes, hyphens, spaces, and periods.
/// Rejects digits and symbols (emails, phone numbers, URLs).
Validator displayName() {
  final re = RegExp(r"^[\p{L}\p{M}' \-\.]+$", unicode: true);
  return all([
    required(),
    (v) => (v != null && v.trim().isNotEmpty && !re.hasMatch(v.trim()))
        ? 'letters, spaces, hyphens, and apostrophes only'
        : null,
    maxLength(64),
  ]);
}

/// Optional display name: skips format check if empty, same rules if provided.
Validator optionalDisplayName() {
  final re = RegExp(r"^[\p{L}\p{M}' \-\.]+$", unicode: true);
  return all([
    (v) => (v != null && v.trim().isNotEmpty && !re.hasMatch(v.trim()))
        ? 'letters, spaces, hyphens, and apostrophes only'
        : null,
    maxLength(64),
  ]);
}

/// Optional email: skips if empty, validates format if provided.
Validator optionalEmail() {
  final re = RegExp(
    r'^[a-zA-Z0-9.!#$%&*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(\.[a-zA-Z0-9-]+)*\.[a-zA-Z]{2,}$',
  );
  return (v) {
    if (v == null || v.trim().isEmpty) return null;
    if (!re.hasMatch(v.trim())) return 'enter a valid email address';
    return null;
  };
}

/// Optional URL: skips if empty, validates scheme + authority + non-bare path if provided.
/// Pass [httpsOnly] to reject non-https URLs.
/// Pass [requirePath] to reject bare domains (no meaningful path).
Validator optionalUrl({bool httpsOnly = false, bool requirePath = false}) {
  return (v) {
    if (v == null || v.trim().isEmpty) return null;
    final s = v.trim();
    final normalized = s.startsWith('http') ? s : 'https://$s';
    final uri = Uri.tryParse(normalized);
    if (uri == null || !uri.hasAuthority) return 'enter a valid URL';
    if (httpsOnly && uri.scheme != 'https') return 'URL must use https';
    if (requirePath && uri.path.replaceAll('/', '').isEmpty) {
      return 'link must point to a specific page, not just the domain';
    }
    return null;
  };
}

/// Password strength: 12+ chars, uppercase letter, number, special character.
Validator password() {
  return all([
    required(),
    (v) => (v != null && v.length < 12) ? 'at least 12 characters' : null,
    (v) => (v != null && v.isNotEmpty && !RegExp(r'[A-Z]').hasMatch(v))
        ? 'include an uppercase letter'
        : null,
    (v) => (v != null && v.isNotEmpty && !RegExp(r'[0-9]').hasMatch(v))
        ? 'include a number'
        : null,
    (v) => (v != null && v.isNotEmpty && !RegExp(r'[^A-Za-z0-9]').hasMatch(v))
        ? 'include a special character'
        : null,
  ]);
}

/// Role name: required, alphanumeric + underscores/hyphens, max 50 chars.
Validator roleName() {
  final re = RegExp(r'^[a-zA-Z0-9_\-]+$');
  return all([
    required(),
    (v) => (v != null && v.trim().isNotEmpty && !re.hasMatch(v.trim()))
        ? 'Letters, numbers, underscores and hyphens only'
        : null,
    maxLength(50),
  ]);
}

File: frontend/src/utils/validators.ts
const displayNameRe = /^[\p{L}\p{M}' .-]+$/u;

export function displayName(value: string | null | undefined): string | null {
  if (!value || value.trim() === '') {
    return 'Required';
  }
  if (!displayNameRe.test(value.trim())) {
    return 'letters, spaces, hyphens, and apostrophes only';
  }
  if (value.trim().length > 64) {
    return 'Max 64 characters';
  }
  return null;
}

export function optionalDisplayName(value: string | null | undefined): string | null {
  if (!value || value.trim() === '') {
    return null;
  }
  if (!displayNameRe.test(value.trim())) {
    return 'letters, spaces, hyphens, and apostrophes only';
  }
  if (value.trim().length > 64) {
    return 'Max 64 characters';
  }
  return null;
}

const emailRe = /^[a-zA-Z0-9.!#$%&*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(\.[a-zA-Z0-9-]+)*\.[a-zA-Z]{2,}$/;

export function optionalEmail(value: string | null | undefined): string | null {
  if (!value || value.trim() === '') {
    return null;
  }
  if (!emailRe.test(value.trim())) {
    return 'enter a valid email address';
  }
  return null;
}

export function optionalUrl(
  value: string | null | undefined,
  options?: { httpsOnly?: boolean; requirePath?: boolean },
): string | null {
  if (!value || value.trim() === '') {
    return null;
  }
  const s = value.trim();
  const normalized = s.startsWith('http') ? s : `https://${s}`;
  try {
    const url = new globalThis.URL(normalized);
    if (!url.hostname) {
      return 'enter a valid URL';
    }
    if (options?.httpsOnly && url.protocol !== 'https:') {
      return 'URL must use https';
    }
    if (options?.requirePath && url.pathname.replaceAll('/', '').length === 0) {
      return 'link must point to a specific page, not just the domain';
    }
    return null;
  } catch {
    return 'enter a valid URL';
  }
}

export function password(value: string | null | undefined): string | null {
  if (!value || value.trim() === '') {
    return 'Required';
  }
  if (value.length < 12) {
    return 'at least 12 characters';
  }
  if (!/[A-Z]/.test(value)) {
    return 'include an uppercase letter';
  }
  if (!/[0-9]/.test(value)) {
    return 'include a number';
  }
  if (!/[^A-Za-z0-9]/.test(value)) {
    return 'include a special character';
  }
  return null;
}

const roleNameRe = /^[a-zA-Z0-9_-]+$/;

export function roleName(value: string | null | undefined): string | null {
  if (!value || value.trim() === '') {
    return 'Required';
  }
  if (!roleNameRe.test(value.trim())) {
    return 'Letters, numbers, underscores and hyphens only';
  }
  if (value.trim().length > 50) {
    return 'Max 50 characters';
  }
  return null;
}

export function required(value: string | null | undefined): string | null {
  if (!value || value.trim() === '') {
    return 'Required';
  }
  return null;
}

export function maxLength(max: number) {
  return (value: string | null | undefined): string | null => {
    if (value && value.trim().length > max) {
      return `Max ${String(max)} characters`;
    }
    return null;
  };
}

export function minLength(
  min: number,
  message?: string,
): (value: string | null | undefined) => string | null {
  return (value: string | null | undefined): string | null => {
    if (!value || value.trim() === '') {
      return null;
    }
    if (value.trim().length < min) {
      return message ?? `At least ${String(min)} characters required`;
    }
    return null;
  };
}
